### ISSUE TITLE
Quantitative Problem

### ISSUE THREAD

**Issue by muzishen** [ankanbhunia/PIDM#37]
https://github.com/ankanbhunia/PIDM/blob/6a9f08c72cd5283bff6a5cfb79875268b6d4ad03/utils/metrics.py#L644
Thank you very much for open-sourcing your articles. 
Could you please clarify what these three files represent and how they were obtained?

**Comment by muzishen** [comment]
And during the dataset preparation, I didn't find the step mentioned 'cd scripts ./download_dataset.sh 'in the readme, which is why I directly downloaded the preprocessed data. Can you help me?  Thank you.

**Comment by ankanbhunia** [comment]
Here the ```real_path``` contains all the training images resized to 256x256 dimension. The individual filename remains the same. Please note while creating this folder - 1. resize using Image.BICUBIC in the PIL library and 2. save images in .png (lossless format)

The folder structure:
```
train_256x256_pngs/
     deepfashionMENDenimid0000008001_1front.png
     deepfashionMENDenimid0000008001_2side.png
     .....
```

**Comment by ankanbhunia** [comment]
It refers to the generated images. You can download our results ([PIDM.zip](https://drive.google.com/file/d/1zcyTF37UrOmUqtRwwq1kgkyxnNX3oaQN/view?usp=sharing)) to undestand how this folder should look like.

**Comment by muzishen** [comment]
> It refers to the generated images. You can download our results ([PIDM.zip](https://drive.google.com/file/d/1zcyTF37UrOmUqtRwwq1kgkyxnNX3oaQN/view?usp=sharing)) to undestand how this folder should look like.

Thank you.
I apologize for the interruption again. 

I don't understand why FID (Fréchet Inception Distance) is computed using the training set and generated images instead of the test set and generated images.

**Comment by ankanbhunia** [comment]
It's a common practice to measure the FID between the training set and generated set when quantifying image synthesis problems. FID doesn't tell you how well the model can reconstruct the images, for that you need to use LPIPS and SSIM. What FID really does is measure how well the model can generate images that closely match the quality of the ones used for training.

**Comment by muzishen** [comment]
Thank you very much.

**Comment by muzishen** [comment]
> It's a common practice to measure the FID between the training set and generated set when quantifying image synthesis problems. FID doesn't tell you how well the model can reconstruct the images, for that you need to use LPIPS and SSIM. What FID really does is measure how well the model can generate images that closely match the quality of the ones used for training.


I apologize for disturbing you again. 
1. According to the synthesized results based on the PIDM you provided, it appears that the FID results cannot be aligned. 
The results are as follows:
![图片](https://github.com/ankanbhunia/PIDM/assets/36322670/87aed909-0296-44d4-9ddc-dfd238e9f753)

It seems a bit strange with the FID results: I first resized it to (256, 256) using PIL with BICUBIC interpolation.

2. Additionally, how can I generate images for PIDM.zip? Could you please provide a script if possible?

**Comment by muzishen** [comment]
`def resize_imgs(images,size):
    img = Image.open(images)
    img = img.convert('RGB')
    resized_img = img.resize(size, resample=Image.BICUBIC)
    return resized_img`
real_path: 37016

**Comment by muzishen** [comment]
Sure, I will give it a try and let you know the results as soon as possible.

**Comment by trungpx** [comment]
> Yes, that is true. But we evaluated it on the entire training set, similar to baselines. 
> 
> Also, I wanted to understand why you are getting such a low FID. So, could you please try it on the 48K images and let me know if you are able to get a similar FID?

I have created the training images from **train_pairs.txt**, removing all duplicated ones, resulting in 37016 images in "**train_256x256_pngs**" folder. I had tried and got the same score of FID:
![image](https://github.com/ankanbhunia/PIDM/assets/37993448/c93fbbe0-e2b4-4e5d-a5c4-7ab5eb2a3a92)

Next, I tried 48674 images as the author suggested, and got the FID = 6.94, not same as paper. I save images as png with: 
    image = Image.open(source_real_filename)
    new_image = image.resize((256, 256), resample=PIL.Image.BICUBIC)
    new_image.save(real_filename)


![image](https://github.com/ankanbhunia/PIDM/assets/37993448/1299c442-7539-4897-864d-e264a9f83f6d)

**Comment by ankanbhunia** [comment]
@trungpx  Please try with the ".png" format as it's considered as a lossless format. jpeg compression can result a significant difference as compared to .png (https://github.com/GaParmar/clean-fid)

**Comment by trungpx** [comment]
@ankanbhunia thanks for the useful repo. Finally, I have successfully replicated the results.
![image](https://github.com/ankanbhunia/PIDM/assets/37993448/6844ba47-5f38-4aef-88f2-8732aa2c677e)

Regrading the question from @muzishen I am also concerning about it: "**how can I generate images for PIDM.zip? Could you please provide a script if possible?**", it is nice if it can be suggested so that I can make some baselines to my research. Thank you.

**Comment by ankanbhunia** [comment]
Thanks for the confirmation.  
Please check the ```gen.py``` inside the  ```utils``` folder. 
Best of luck!

**Comment by muzishen** [comment]
> 

Hello, I have confirmed that there are 48,674 images. 
They are in PNG format. 
The results are as follows: FID: 6.3798; SPIPS: 0.1678; SSIM: 0.7320.

**Comment by muzishen** [comment]
> > Yes, that is true. But we evaluated it on the entire training set, similar to baselines.
> > Also, I wanted to understand why you are getting such a low FID. So, could you please try it on the 48K images and let me know if you are able to get a similar FID?
> 
> I have created the training images from **train_pairs.txt**, removing all duplicated ones, resulting in 37016 images in "**train_256x256_pngs**" folder. I had tried and got the same problematic score of FID as @muzishen: ![image](https://user-images.githubusercontent.com/37993448/246185109-c93fbbe0-e2b4-4e5d-a5c4-7ab5eb2a3a92.png)
> 
> Next, I tried 48674 images as the author suggested (**train.lst**), and got the FID = 6.94, not really same as paper. I save images as jpg with:
> 
>     * ```
>       image = Image.open(source_real_filename)
>       ```
> 
>     * ```
>       new_image = image.resize((256, 256), resample=PIL.Image.BICUBIC)
>       ```
> 
>     * ```
>       new_image.save(real_filename)
>       ```
> 
> 
> The generated output images I got from PIDM.zip provided by the author. Here is my results :(
> 
> ![image](https://user-images.githubusercontent.com/37993448/246200311-1299c442-7539-4897-864d-e264a9f83f6d.png)
> 
> Could you confirm the exact number of training images in your PC is 48,674, too?
> 
> I'm trying again with "png" and update soon. [updated below comments]

Hello, I have confirmed that there are 48,674 images. 
They are in PNG format. 
The results are as follows: FID: 6.3798; SPIPS: 0.1678; SSIM: 0.7320.

**Comment by muzishen** [comment]
@trungpx  Hello, I have confirmed that there are 48,674 images. 
They are in PNG format. 
The results are as follows: FID: 6.3798; SPIPS: 0.1678; SSIM: 0.7320.

**Comment by trungpx** [comment]
> @ankanbhunia @muzishen @trungpx Thanks for your questions, I try to reproduce the experimental results but fail to it. I would like to ask you if my procedures are correct.
> 
> 1. Split the deepfashion/img dataset to train(48674) and test(4038) folder with .jpg format.
> 2. Crop images in PIL center crop with code img.crop((40, 0, 216, 256)) and then use PIL to resize images with code image.resize((256, 256), resample=PIL.Image.BICUBIC) for train and test sets.
> 3. The real_path is 48674 train images produced by previous steps, gt_path is 4038 test images and distorated_path is PIDM images(8570).
> 
> But my result is FID: 8.7594, LPIPS: 0.2020, ssim_256: 0.714411.
> 
> Do you know which step I did was not correct? Thanks for your help

The second step might be problem, dont use anh crop, just resize it as you did. Try and let me know if it fixed.

**Comment by trungpx** [comment]
Could you try again with the following:
1) split train (48k with **train.lst**) and test (3144 with test_pair.txt), open .jpg and resize and save to .png, example as below in my case:
    source_real_filename = './data/deepfashion/img/' + file
    image = Image.open(source_real_filename)
    new_image = image.resize((256, 256), resample=PIL.Image.BICUBIC)
    real_filename = real_path + s1.replace('.jpg','') + '.png'
    new_image.save(real_filename)
2) use the following list for test images:
filenames_test = []
file_txt_test = '{}/test_pairs.txt'.format(root)
with open(file_txt_test, 'r') as f:
    lines = f.readlines()
    for item in lines:
        filenames_test.extend(item.strip().split(','))                
filenames_test = list(set(filenames_test)) # len = 3144
And then open with PIL for these **jpg** test img and resize, save to **png**.

root2 = './data/deepfashion/'
for files in tqdm(filenames_test):
    s1=os.path.splitext(files)[0]
    # print(s1)
    s2 = s1.replace('/','')
    # print(s2)
    s3 = s2.replace('id_','id')
    # print(s3)
    # print(s3.rfind('_'))
    s4 = s3[:s3.rfind('_')] + s3[s3.rfind('_')+1:]
    s5 = 'deepfashion' + s4[3:] + '.png'
    # print(s5)
    real_filename = gt_path + s5
    # print(real_filename)
    source_real_filename = root2 + files
    # print(source_real_filename)
    # shutil.copyfile(source_real_filename,real_filename)
    image = Image.open(source_real_filename)
    new_image = image.resize((256, 256), resample=PIL.Image.BICUBIC)
    # new_image = image.resize((256, 256))
    new_image.save(real_filename)

![image](https://github.com/ankanbhunia/PIDM/assets/37993448/46da4b40-aa4b-4667-989a-fbe65b30ccdb)

It should solve the problem.

**Comment by nicolasugrinovic** [comment]
Hi to all,
I have a question regarding the number of training pairs. In the paper it is mentioned "Following the same data configuration in [30], we split this dataset into **training** and testing subsets with **101,966** and 8,570 pairs". However, the train_pairs.txt contains **~37K** pairs as you mention. Where does this discrepancy of the number of pairs come from?
Thank you!

**Comment by trungpx** [comment]
> Hi to all, I have a question regarding the number of training pairs. In the paper it is mentioned "Following the same data configuration in [30], we split this dataset into **training** and testing subsets with **101,966** and 8,570 pairs". However, the train_pairs.txt contains **~37K** pairs as you mention. Where does this discrepancy of the number of pairs come from? Thank you!

Hello,

You can see in the content of the "**train_pairs.txt**" that each line contains multiple file names separated by a comma, which means that 37,016 lines are not only 37k pairs. But the total should be around 101,966 pairs. Hope it is clear.

![image](https://github.com/ankanbhunia/PIDM/assets/37993448/7e677467-08c6-4c16-9ccc-bd5ce22264ef)
